### PR TITLE
Do not report `AddedRequiredProperty` if the property is `readOnly`

### PR DESCRIPTION
**Pull request by konrad-jamrozik**
Addresses:

- https://github.com/Azure/azure-sdk-tools/issues/7184

Per this specification:

- https://github.com/Azure/azure-sdk-tools/issues/7184#issuecomment-2162237789

### CODE REVIEW

**Comment by konrad-jamrozik** [comment]
@mikekistler in this PR I am addressing:

https://github.com/Azure/openapi-diff/blob/master/docs/rules/1034.md

But there are also these rules:

https://github.com/Azure/openapi-diff/blob/master/docs/rules/1010.md
https://github.com/Azure/openapi-diff/blob/master/docs/rules/1025.md

Do they also need to be adjusted?

File: openapi-diff/src/modeler/AutoRest.Swagger.Tests/Resource/Swagger/old/added_readonly_required_property.json [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+          "type": "string",
+          "description": "Enum.",
+          "enum": [ "a1", "a2", "a3" ]
+
+        },
+        "b": {
+          "type": "string",
+          "description": "Enum.",
+          "enum": [ "b1" ],
+          "readOnly": true
```
Comment (konrad-jamrozik) [inline openapi-diff/src/modeler/AutoRest.Swagger.Tests/Resource/Swagger/old/added_readonly_required_property.json:44]: FYI this `readOnly` here is the key difference. All the rest was copied over from `added_required_property.json`. Same for `new` variant.

File: openapi-diff/src/modeler/AutoRest.Swagger/Model/Schema.cs [hunk 2]
```diff
@@ -184,27 +184,58 @@ Schema previous
...
 
-            if (Required != null && priorSchema.Required == null)
+            var newRequiredNonReadOnlyPropNames = new List<string>();
+            foreach (string requiredPropName in Required)
             {
-                context.LogBreakingChange(ComparisonMessages.AddedRequiredProperty, String.Join(", ", Required));
-                return;
+                Properties.TryGetValue(requiredPropName, out Schema propSchema);
+                priorSchema.Properties.TryGetValue(requiredPropName, out Schema priorPropSchema);
+                if (propSchema is null || priorPropSchema is null)
```
Comment (konrad-jamrozik) [inline openapi-diff/src/modeler/AutoRest.Swagger/Model/Schema.cs:215]: @mikekistler due to this check:

`priorPropSchema is null`

this rule will flag on any properties that are `required` but not actually defined, even if there were no changes to them.

Let me know if you don't want that.

Comment (mikekistler) [reply openapi-diff/src/modeler/AutoRest.Swagger/Model/Schema.cs:215]: It is legal (but very uncommon) to name properties in `required` that are not defined properties in the schema.

I think the desired behavior is to only flag properties when then changed from not-required to required, regardless of whether they are defined properties.

**Review by mikekistler — APPROVED** [review]
Looks good. 👍 

With the qualifications that I noted.

File: openapi-diff/src/modeler/AutoRest.Swagger/Model/Schema.cs (hunk 2 above)
Comment (konrad-jamrozik) [reply openapi-diff/src/modeler/AutoRest.Swagger/Model/Schema.cs:215]: @mikekistler I will adjust as follows:

1. If the property **was** required and **is** required: do not report a breaking change. This is regardless of property being actually defined or not.
2. If the property **was not** required and **is** required then:
  - If the property **was defined and read-only** and **is defined and read-only** then do not report breaking change.
  - In all other cases, report breaking changes.

**Comment by konrad-jamrozik** [comment]
[Pull Request 559774](https://devdiv.visualstudio.com/DevDiv/_git/openapi-alps/pullrequest/559774): Update `@azure/oad` to `0.10.11` + `rush update`

**Comment by konrad-jamrozik** [comment]
Deployed to production:
https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=9767803&view=results

**Pull request merged**